File: python/2402.meeting-rooms-iii.py
import heapq


class Solution:
    def mostBooked(self, n: int, meetings: list[list[int]]) -> int:
        busy = []  # (end, room)
        free = list(range(n))  # (room)
        rooms = [0] * n

        meetings.sort()

        for start, end in meetings:
            # free meetings that is over
            while busy and busy[0][0] <= start:
                _, room = heapq.heappop(busy)
                heapq.heappush(free, room)

            # no delay
            if free:
                room = heapq.heappop(free)
                heapq.heappush(busy, (end, room))
            else:
                ending, room = busy[0]
                duration = end - start
                heapq.heapreplace(busy, (ending + duration, room))

            rooms[room] += 1

        # argmax
        return rooms.index(max(rooms))

File: cpp/2402.meeting-rooms-iii.cpp
#include <queue>
#include <utility>
#include <vector>

using std::vector, std::priority_queue, std::pair, std::greater;

class Solution {
public:
  int mostBooked(int n, vector<vector<int>> &meetings) {
    vector<int> res(n, 0);

    sort(meetings.begin(), meetings.end(),
         [](const auto &p1, const auto &p2) -> bool {
           return p1[0] == p2[0] ? p1[1] < p2[1] : p1[0] < p2[0];
         });

    priority_queue<int, vector<int>, greater<int>> free;
    for (int i = 0; i < n; i++) {
      free.push(i);
    }

    priority_queue<pair<long, int>, vector<pair<long, int>>,
                   greater<pair<long, int>>>
        busy;

    for (const auto &meeting : meetings) {
      int start = meeting[0];
      int end = meeting[1];

      while (!busy.empty() && busy.top().first <= start) {
        free.push(busy.top().second);
        busy.pop();
      }

      if (!free.empty()) { // there is free meeting rooms
        int room = free.top();
        free.pop();
        busy.push({end, room});
        res[room]++;
      } else { // there is no free meeting rooms, postpone meeting
        int room = busy.top().second;
        long diff = busy.top().first - start;
        busy.pop();
        busy.push({diff + end, room});
        res[room]++;
      }
    }

    int max_meetins = 0;
    for (int i = 0; i < n; i++) {
      if (res[i] > res[max_meetins]) {
        max_meetins = i;
      }
    }

    return max_meetins;
  }
};